[docs] Add CONTRIBUTING.md

anik120 opened the pull request:
Include instructions for setting up a dev environment,
conventions on commit message formatting, and other 
resources to help get contributions into marketplace-operator.

anik120 commented:
/cc @awgreene @SamiSousa @galletti94 @jeremylinlin @tkashem

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,62 @@
+
+- Email: [operator-framework](https://groups.google.com/forum/#!forum/operator-framework)  
+
+## Setting up a dev environment
```
Marketplace also works on vanilla Kubernetes. Should we at least mention that here? Just that, this is an example environment and vanilla kube and `kubectl` should work in addition to `oc` ?

anik120 replied:
@kevinrizza do we have a doc detailing steps that can be taken to get marketplace-operator running on vanilla Kubernetes (installing olm -> marketplace-operator)? I was thinking about writing something along the lines of:
```
This section provides detailed instructions on how to set up a development environment on an Openshift cluster. To use maketplace-operator in a vanilla Kubernetes cluster instead, follow the instructions [here] (<link-to-k8s-installation-doc>)  and use the `kubectl`/`oc` command. 
```
But I couldn't find any in the docs folder.

aravindhp replied:
>But I couldn't find any in the docs folder.

Isn't that what you are writing :-)

This definitely needs an OpenShift vs plain vanilla Kuberenetes sections to call out the differences.

kevinrizza replied:
We don't have a doc, but the process should be identical except using `kubectl` instead of `oc`. The operator-sdk works with either.

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,62 @@
+
+this uses tmux to setup a test cluster that can easily be killed and started for debugging.
+
+Fixes #38
```
This section seems irrelevant to `operator-marketplace`. What is # 38 and why are we referencing tmux? and a test-cluster command?

anik120 replied:
@kevinrizza this is suppose to be an example commit message. I've updated this section with a commit message from our repo to make it more relevant.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+## Setting up a dev environment
+
+- Fork the repository on GitHub
+- Clone the repository in your go path.
```
`forked repository`

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+- Fork the repository on GitHub
+- Clone the repository in your go path.
+- Get access to an Openshift cluster. The marketplace-operator is currently supported on Openshift 4.0. You can find instructions on installing an Openshift cluster [here](https://github.com/openshift/installer).
+- Turn off the Cluster-Service-Operator (CVO). The CVO manages the lifecycle of the marketplace-operator along with other internal Openshift components. Once the CVO is truned off, the marketplace-operator deployment can be deleted. 
```
Please don't turn off CVO. Instead provide a `ClusterVersion` with `overrides`.
```
apiVersion: config.openshift.io/v1
kind: ClusterVersion
metadata:
  name: version
spec:
  channel: fast
  upstream: https://api.openshift.com/api/upgrades_info/v1/graph
  overrides:
    - kind: Deployment
      name: marketplace-operator
      namespace: openshift-marketplace
      unmanaged: true
```

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+- Delete the marketplace-operator deployment `oc delete deployment openshift-marketplace -n marketplace-operator`
+- Install [Operator-SDK](https://github.com/operator-framework/operator-sdk).
+- `cd` into the marketplace-operator repository and `dep ensure` to download all dependencies.
+- `operator-sdk up local --namespace=openshift-marketplace --kubeconfig=<path-to-kubeconfig-file>` to complie the marketplace-operator and start the operator in you dev environment.      
```
Doing all the testing in this manner will not test RBAC changes if that is being made. For that you also need instructions on how to use your own `Deployment` where you have replaced the image with the one you are building using the SDK.

kevinrizza replied:
+1

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+    `oc scale --replicas 0 -n openshift-cluster-version deployments/cluster-version-operator`
+- Delete the marketplace-operator deployment `oc delete deployment openshift-marketplace -n marketplace-operator`
+- Install [Operator-SDK](https://github.com/operator-framework/operator-sdk).
+- `cd` into the marketplace-operator repository and `dep ensure` to download all dependencies.
```
Why is this needed? We should have everything in the `vendor` folder.

kevinrizza replied:
@aravindhp to build yes, but for whatever reason the sdk does not read those dependencies locally. at least that is what we found on friday when trying to get this to run in Anik's local environment.

aravindhp replied:
That should be called out here. This should also be brought up with the SDK team.

anik120 replied:
@kevinrizza I think that might have been due to the fact that my cloned repo was not in my go path.

kevinrizza replied:
@anik120 did you confirm with a clean environment?

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+## Reporting bugs and creating issues
+
+Reporting bugs is one of the best ways to contribute. If any part of the operator-marketplace project has bugs or documentation mistakes, please let us know by [opening an issue](https://github.com/operator-framework/operator-marketplace/issues/new)
```
I suggest leaving out `Reporting bugs is one of the best ways to contribute.`

`opening an issue or a PR`

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+This is a rough outline of what a contributor's workflow looks like:
+
+- Create a topic branch from where to base the contribution. This is usually master.
```
Mention that if you want to sync up with the upstream master, that a rebase workflow must be followed.

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+Operator Marketplace is Apache 2.0 licensed and accepts contributions via GitHub pull requests. This document outlines some of the conventions on commit message formatting, instructions on how to set up a dev environment, and other resources to help get contributions into operator-marketplace.
+
+# Email and Chat
```
There's no email listed here under the Email header. Additionally, that link is to the operator-framework google group which we don't control and afaik even look at. We don't have a public email list set up either, so is this header necessary? Maybe we should set up something and add it later?

aravindhp replied:
That is a good point. Main way to reach out should be issues I guess.

aravindhp commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,63 @@
+<footer>
+```
+
+The first line is the subject and should be no longer than 50 characters, the second line is always blank, and other lines should be wrapped at 72 characters. This allows the message to be easier to read on GitHub as well as in various git tools.
````
Are you following the 72 character limit in your commit message? :-)

anik120 replied:
@aravindhp For the other PRs yes. For this one I edited my message now and it's following the convention :)

aravindhp replied:
We can go with 80 characters instead of 72.

anik120 commented:
@kevinrizza @aravindhp addressed your comments. Discussed testing changes to RBAC with Kevin and penned down a few lines about that. Let me know if those lines are missing anything/not correct.

kevinrizza commented on examples/cvo.override.yaml:
```diff
@@ -0,0 +1,12 @@
+    - kind: Deployment
+      name: marketplace-operator
+      namespace: openshift-marketplace
+      unmanaged: true
```
Missing newline at the end of the file.

aravindhp replied:
Please add a newline.

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+## Marketplace-operator on vanilla Kubernetes
+
+- To set up marketplace-operator on vanilla Kubernetes, [install Operator-Lifecycle-Manager(OLM)](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/install/install.md) in a Kubernetes Cluster. 
+- `cd` into the makertplace-operator repository and install the marketplace-operator using `kubectl apply -f deploy/upstream`.
```
should be marketplace-operator

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- `operator-sdk up local --namespace=openshift-marketplace --kubeconfig=<path-to-kubeconfig-file>` to complie the marketplace-operator and start the operator in your dev environment. 
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a seperate deployment. Easiest way to do this is to do a local `docker build` and then push the image to a registry. Replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/operator.yaml#L23) with the newly built image, then do a `oc apply -f deploy/`  ...
```
`separate`

aravindhp replied:
`oc apply -f deploy/operator.yaml`

aravindhp replied:
Why `docker build` vs `operator sdk build`?

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a seperate deployment. Easiest way to do this is to do a local `docker build` and then push the image to a registry. Replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/operator.yaml#L23) with the newly built image, then do a `oc apply -f deploy/`  ...
+
```
Can you specifically call out to use the `Dockerfile` in the root of the repo here? There are several dockerfiles in the project so it could be confusing.

kevinrizza commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- Fork the repository on GitHub
+- Clone the forked repository in your go path.
+- Get access to an Openshift cluster. The marketplace-operator is currently supported on Openshift 4.0. You can find instructions on installing an Openshift cluster [here](https://github.com/openshift/installer).
+- The Cluster-Version-Operator(CVO) manages the lifecycle of the marketplace-operator along with other internal Openshift components. To delete the marketplace-operator deployment, insruct CVO to stop managing the marketplace-operator by applying the configuration file that can be found [here](examples/cvo.override.yaml). 
```
`To delete the marketplace-operator deployment` makes it sound like adding the override deletes the deployment. Can it be something like `Before deleting the marketplace-operator deployment, instruct CVO to ....(etc)` ?

aravindhp replied:
Please add the command to apply the override given you are doing this everywhere else.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- To delete the marketplace-operator deployment use `oc delete deployment marketplace`. 
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a seperate deployment. Easiest way to do this is to do a local `docker build` and then push the image to a registry. Replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/upstream/07_operator.yaml#L28) with the newly built image, then do a `kubectl a ...
```
Same comment as above.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+
+## Contribution flow
+
+This is a rough outline of what a contributor's workflow looks like:
```
Please remove rough

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+
+### Format of the commit message
+
+We follow a rough convention for commit messages that is designed to answer two
```
Please remove rough

anik120 commented:
@kevinrizza @aravindhp addressed all comments.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- `operator-sdk up local --namespace=openshift-marketplace --kubeconfig=<path-to-kubeconfig-file>` to complie the marketplace-operator and start the operator in your dev environment. 
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a separate deployment. Easiest way to do this is to do a local `operator-sdk build` using [this dockerfile](./Dockerfile) and then `docker push` the image to a registry. Replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/operator.yaml#L23) with th ...
```
Please add a section using ``` listing the workflow. I find it hard to follow by reading all the text. Moreover doesn't `operator-sdk build` pick the `Dockerfile`? So you just need to mention that it ends up using that particular `Dockerfile`.

anik120 replied:
@aravindhp restructured the document. Let me know what you think now.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+This is an outline of what a contributor's workflow looks like:
+
+- Create a topic branch from where to base the contribution. This is usually master.
+- Make commits of logical units.
```
This needs to be expanded. Please mention that this means a commit should typically try to do only one thing. For example fixing a bug and adding a feature shouldn't be in the same commit. Mixing bugs and feature shouldn't be done in the same PR for that. A person's individual workflow adding a change shouldn't be reflected in the commits.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- Create a topic branch from where to base the contribution. This is usually master.
+- Make commits of logical units.
+- Make sure commit messages are in the proper format (see below).
+- Push changes in a topic branch to a personal fork of the repository.
```
Shouldn't the first step be to for the repo?

anik120 replied:
@aravindhp could you repeat this.

aravindhp replied:
Sorry, shouldn't forking the repo be the first step?

anik120 replied:
@aravindhp I think I have listed forking the repo as the first step. Please let me know if you can see otherwise.

aravindhp replied:
You have in `Setting up a dev environment`. Here you just need to mention that you are creating a topic branch in the forked repo in your first bullet.

anik120 replied:
@aravindhp Ah right. Added that now.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,70 @@
+- Push changes in a topic branch to a personal fork of the repository.
+- To make sure that your topic branch is in sync with the remote master branch, follow a rebase workflow.
+- Submit a pull request to operator-framework/operator-marketplace.
+- The PR must receive a LGTM from two maintainers of the operator-marketplace project.
```
It is one /lgtm and one /approve from the maintainers. We try to give two /lgtm but there is nothing in the CI process that looks for that.

aravindhp commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,117 @@
+$ oc delete deployment openshift-marketplace -n marketplace-operator
+```
+- Install [Operator-SDK](https://github.com/operator-framework/operator-sdk).
+- Complie the marketplace-operator and start the operator in your dev environment
````
`Compile`

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,117 @@
+- Fork the repository on GitHub
+- Clone the forked repository in your go path.
+- Get access to an Openshift cluster. The marketplace-operator is currently supported on Openshift 4.0. You can find instructions on installing an Openshift cluster [here](https://github.com/openshift/installer).
+- The Cluster-Version-Operator(CVO) manages the lifecycle of the marketplace-operator along with other internal Openshift components. Before deleting the marketplace-operator deployment, insruct CVO to stop managing the marketplace-operator 
```
`instruct CVO`

aravindhp commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,117 @@
+```
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a separate deployment. Easiest way to do this is to build an image locally using [this dockerfile](./Dockerfile), push the image to a registry, and replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/operator.yaml#L23) with the newly built image. T ...
````
`policies`

aravindhp commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,117 @@
+``` 
+
+### Testing changes to RBAC policy
+- If there's any modification in existing RBAC poilicies, the modifications need to be tested using a seperate deployment. Easiest way to do this is to build an image locally using [this dockerfile](./Dockerfile), push the image to a registry, and replace the [image here](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/upstream/07_operator.yaml#L28) with the newly bu ...
````
`policies`
`separate`

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,117 @@
+
+$ docker push <REGISTRYHOST>/<USERNAME>/marketplace-operator 
+
+$ vi deploy/upstream/07_operator.yaml
```
Please use a `sed` here to do the replacement

anik120 commented:
@aravindhp I think it's safe to say all the spell-checking tools out there ruined my typing skills. Corrected all the errors. Also added instructions to use sed to replace images.

aravindhp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,109 @@
+
+$ docker push <REGISTRYHOST>/<USERNAME>/marketplace-operator 
+
+$ sed -i -e 's?quay.io/openshift/origin-operator-marketplace:latest?quay.io/anik120/marketplace-operator?g' deploy/operator.yaml
```
Please use `<REGISTRYHOST>/<USERNAME>/` like you have in other places.

SamiSousa commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,111 @@
+
+- Fork the repository on GitHub
+- Clone the forked repository in your go path.
+- Get access to an Openshift cluster. The marketplace-operator is currently supported on Openshift 4.0. You can find instructions on installing an Openshift cluster [here](https://github.com/openshift/installer).
```
nit: Should it be `OpenShift` or `Openshift`?

anik120 replied:
@SamiSousa looks like it's OpenShift :)

SamiSousa commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,111 @@
+$ oc delete deployment marketplace
+``` 
+
+### Testing changes to RBAC policy
````
nit: Since this section is specific for testing RBAC on vanilla kube, can the header reflect this? I was scrolling through and thought this was the same as the section for OpenShift, but this is specific for kubernetes. On a quick scroll it's tough to distinguish the two sections.

anik120 replied:
@SamiSousa changed the title. Please let me know if it's easier to distinguish now.

SamiSousa commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,111 @@
+
+$ sed -i -e 's?quay.io/openshift/origin-operator-marketplace:latest?<REGISTRYHOST>/<USERNAME>/marketplace-operator?g' deploy/upstream/07_operator.yaml
+
+$ oc apply -f deploy/upstream/operator.yaml
```
`kubectl apply`

anik120 replied:
@SamiSousa thank you for noticing this. Addressed it.
